### ISSUE TITLE
doc : travis devel example depends on trusty

### ISSUE THREAD

**Issue by asmodehn** [ros-infrastructure/ros_buildfarm#470]
I am talking about https://github.com/ros-infrastructure/ros_buildfarm/blob/master/doc/jobs/devel_jobs.rst#run-the-devel-job-on-travis which is pointed from the wiki.

It would be better to have a docker-based job, to be able to finally move on from trusty...

Example using docker (with custom script instead of buildfarm) : https://github.com/pyros-dev/pyros-utils/blob/0.1.4/.travis.yml

I have no idea what modification would be needed in the buildfarm scripts though (probably not much since the buildfarm already uses docker ?)

**Comment by asmodehn** [comment]
Just noticed the build farm shell script does include the docker calls...
It would be nice if it could integrate directly with the travis build matrix...

**Comment by dirk-thomas** [comment]
Trusty is the latest version supported by Travis. And it is only being used to fetch the Python package and invoke the build. The build itself is already running within Docker and uses whatever distro the job needs.

I am not sure what you are trying to achieve but you should be able to setup a Travis build matrix which builds either different job types and/or for different target platforms. Maybe you can share some more information so that it becomes possible to give you more directed feedback.

**Comment by asmodehn** [comment]
I am currently trying to find a clean and easily maintainable way to get [multi ros/linux distro test in one build](https://travis-ci.org/pyros-dev/pyros-utils/builds/213310387), and even [multi usecase for third party releases](https://travis-ci.org/asmodehn/marshmallow-rosrelease/branches). I currently do that using a custom script and I would rather use a build script maintained by others, but I haven't found anything as simple and clear yet.
I recently tested with [industrial_ci](https://github.com/ros-industrial/industrial_ci) [here](https://travis-ci.org/pyros-dev/pyros-utils/builds/261293561) and the output is similar, but I think too much is hidden in shell scripts, so I'm trying ros_buildfarm scripts as an alternative...

It [seems](https://travis-ci.org/ros-infrastructure/ros_buildfarm/builds/261207028) that the buildfarm script would "hide" the docker details, and make the build a harder to debug. I would rather use docker in the travis file, and be able to test a package for multiple ros/linux distribution because of it, while matching the travis matrix.
Then add a simple shell script to be able to test different usecases locally or "inside the docker", like I am doing currently...

So I guess I am asking "Is there a way to `generate_devel_script.py` without anything related to docker... only ROS package build & test commands ?", that would look like [this](https://github.com/pyros-dev/pyros-utils/blob/0.1.4/travis_checks.bash). That is the part where I want to be as close as possible to the buildfarm process, and leave the docker mess in [.travis.yml](https://github.com/pyros-dev/pyros-utils/blob/0.1.4/.travis.yml).

**Comment by dirk-thomas** [comment]
> I am currently trying to find a clean and easily maintainable way to get multi ros/linux distro test in one build, and even multi usecase for third party releases.

That should be easily doable. For each entry in the build matrix you just need to invoke `generate_devel_script.py` with custom arguments (or use specific environment variables for each matrix item).

> So I guess I am asking "Is there a way to generate_devel_script.py without anything related to docker... only ROS package build & test commands ?"

As far as I understand your question I think the answer is no. The ROS buildfarm is designed to run each job in a clean environment which makes the builds deterministic and reproducible. Every build is always done in Docker to setup the necessary dependencies etc. See the [devel job docs](https://github.com/ros-infrastructure/ros_buildfarm/blob/master/doc/jobs/devel_jobs.rst) as an example.

If you really want to go through the effort of managing the Docker container yourself you might be able to run the `ros_buildfarm` scripts which are commonly executed within Docker manually. Maybe the [graph](https://github.com/ros-infrastructure/ros_buildfarm/blob/master/doc/jobs/devel_call_graph.png) how a devel job works can be helpful to figure out what happens internally and how those could be called directly.

> that would look like this:
> > mkdir -p testbuild
cd testbuild
cmake .. -DCMAKE_INSTALL_PREFIX=./install
if [ "$ROS_FLOW" == "devel" ]; then
    make -j1
    source devel/setup.bash
    make -j1 tests
    make -j1 run_tests
    catkin_test_results .
elif [ "$ROS_FLOW" == "install" ]; then
    make -j1 install
    source install/setup.bash
    python -m nose pyros_utils
fi

The referenced approach can't work in general. E.g. Travis only provides Trusty environments. You couldn't test a ROS Lunar build which requires Xenial, Yakkety, or Zesty. The same for Debian jobs.

Also the referenced snippet only covers a small fraction of what the buildfarm needs to do for every job like installing dependencies with rosdep, separate the build and install phase, etc. Replicating all that in your Travis file would be a lot of effort.

> It seems that the buildfarm script would "hide" the docker details, and make the build a harder to debug.

The usage of a known workflow as defined by `ros_buildfarm` makes your build easily reproducible. On the other hand a custom build adds additional uncertainty since users can't ensure that they can reproduce the problem locally.

On Travis your debugging options are anyway very limited. Therefore I think it is even more important that you can reproduce the exact same build locally since locally you can easily debug each and every aspect of the process (go through the file system, start a Docker container interactively, etc.).

> That is the part where I want to be as close as possible to the buildfarm process, and leave the docker mess in .travis.yml.

Why would you need to deal with any "Docker mess" when setting up Travis using `ros_buildfarm`? All of that is done "behind the scene" for you (see [example](https://github.com/ros-infrastructure/ros_buildfarm/blob/master/doc/jobs/devel_jobs.rst#run-the-devel-job-on-travis)).

**Comment by asmodehn** [comment]
The ease of reproducibility is indeed the main goal here, and I agree that having docker in the script makes the whole build more easily reproducible.

I guess I am trying to keep docker out of the "build" since I see it more as a "CI platform artifact" that does not seem to be part of the software itself, that the neither the user or developer of the software enjoys dealing with the extra complexity it brings.

Another reason might be to integrate with whatever "isolation solution" any CI already provides (docker, or VM, or anything else...) when not using the OSRF buildfarm, even if I want to use the same "build process"... but the lines start to get blurry there.

Thanks for the graph link, it is instructive. I will have a deeper look and check if/how I can call the scripts there directly.

**Comment by dirk-thomas** [comment]
> Thanks for the graph link, it is instructive. I will have a deeper look and check if/how I can call the scripts there directly.

As the graph indicates the two levels of Docker are currently used because the first Docker container contains the necessary dependencies to decide what is needed (e.g. which ROS dependencies) in the second Docker container which does the actual build.

If there is anything in the package which could be refactored to make it easier to use parts of the functionality please feel free to propose suggestions.

Is there anything else which should be done based on this ticket at the moment or should this be closed for now? We can for sure continue the conversation on the ticket.

**Comment by asmodehn** [comment]
I worked on this a bit more and I wasn't sure if I should make another issue or not... anyway here it is : 

I am not sure why the error is triggered there : https://travis-ci.org/pyros-dev/pyros-utils/jobs/270673962

That package is building properly on the buildfarm and it is declared in rosdistro (with a source section) : https://github.com/ros/rosdistro/blob/master/indigo/distribution.yaml#L9689

Also the build seems to be working fine on the OSRF buildfarm...

Is it triggered because some kind of git url matching is taking place and I have moved the repository (asmodehn-> pyros-dev)?
Or is it because I had to rename the repository (since the repository name doesnt match the ros name i rosdistro - `pyros-utils` -> `pyros_utils`) ?

Why should I even rely on rosdistro for this ? the goal is only to build a package, I should not need any rosdistro...

Thanks for any hint you may have... I m still trying to use rosbuildfarm to build some existing (and already released) packages: https://github.com/pyros-dev/pyros-utils/pull/10

**Comment by dirk-thomas** [comment]
In the referenced Travis log you are building for `lunar` and the distribution file for lunar doesn't have a source section: https://github.com/ros/rosdistro/blob/249f54a11ea585d8b53e1ef3573686819468522c/lunar/distribution.yaml#L2041-L2046

**Comment by dirk-thomas** [comment]
> Why should I even rely on rosdistro for this ? the goal is only to build a package, I should not need any rosdistro...

Your invocation passes only a `REPOSITORY_NAME`. It needs `rosdistro` to map that name to a repository.

**Comment by asmodehn** [comment]
Thanks, but let me rephrase that sentence...
Since I only want to build the content of a repository, to quickly run a bunch of tests, before sending a version to (any) buildfarm, I was looking for a way to run more or less the same commands as the buildfarm, with a simpler environment setup. That is the middle step in that workflow : 

local manual checks -> quick automated isolated checks -> full buildfarm build

It may be a bias on my part, since it s the workflow I got used to, developing for python, testing with different python environments (travis), before building the full ros package on the buildfarm.

I realize I also unconsciously applied the same workflow to a normal ros package : 

local build and test -> [travis catkin build and tests (fast)](https://travis-ci.org/pyros-dev/pyros-utils/builds/213310369) -> full buildfarm build (slow)

It could be also because I haven't used the devel job of the buildfarm much for the packages I m working on, I went for travis a while ago already, when the buildfarm was too slow to be usable for this...

So now it seems that the buildfarm scripts do not provide an interface for this kind of small fast builds (no extra rosdistro), correct ? so I can : 
- drop that intermediate step, avoid duplication with the buildfarm job on travis (there is no point having it while the speed is similar)
- keep that intermediate step for quick and easy build tests, but use something else ( my old custom `.travis.yml` or [industrial_ci](https://github.com/pyros-dev/pyros-utils/blob/industrial_ci/.travis.yml) ) which is a bit [closer to what I would expect](https://github.com/pyros-dev/pyros-utils/blob/industrial_ci/.travis.yml). Also in terms of speed : https://travis-ci.org/pyros-dev/pyros-utils/builds/263399132 - I didn't investigate the reason )

Anyway, thanks for the introduction to the buildfarm scripts !

**Comment by dirk-thomas** [comment]
If I understand your "quick automated isolated checks" correctly I think the "problem" is that it doesn't work generically. As soon as your package has dependencies on other ROS packages it needs the "extra overhead" of the rosdistro to resolve these to actual Debian packages.

If you don't want to resolve the repository name through rosdistro you could instead pass the repository information explicitly (as describe for prerelease jobs: https://github.com/ros-infrastructure/ros_buildfarm/blob/master/doc/jobs/prerelease_jobs.rst#run-for-custom-repositories). If your package doesn't have any dependencies on other ROS packages the rosdistro you pass shouldn't make any difference. That will at least allow you to run a CI job without the need for the repo to be registered in the rosdistro.

**Comment by asmodehn** [comment]
Yes I think it s just a mismatching of interface provided by that script, between what is there and what I'd expect... but that is fine, I might not be the intended usecase.

From what I understand I need rosdistro `distribution.yaml` even for the _current_ package I am building (which travis or any other CI system will have got already via their own integrated way). And I can't make [that](https://travis-ci.org/pyros-dev/pyros-utils/jobs/272821623) [work](https://travis-ci.org/pyros-dev/pyros-utils/jobs/272818598) just yet...

For the dependencies, relying on rosdep (via rosdistro) makes sense, to get the debs as I usually do locally, but not relying on the `distribution .yaml`, since I might want to do something without/before declaring it on rosdistro (testing doc build, testing devel build, or testing different version etc.)...

**Comment by dirk-thomas** [comment]
As said you can use a `prerelease` job to avoid the need for your repo to be part of the distribution file. If you have the repo already cloned you can place it in the same location where the job would clone it to. The script will "warn" you about that and than use the existing folder I think.
